File: src/main/sync/providers.ts
import { FerroError } from '@shared/errors'
import { SYNC_FILE_NAME } from '@shared/sync'
import { createLogger } from '../core/logger'

const log = createLogger('sync')

// ── Uzak depo sağlayıcıları ────────────────────────────────────────────────
// Her sağlayıcı yalnızca ŞİFRELİ blob'u taşır; içerik hakkında bilgisi yoktur.
// download() null dönerse uzak dosya henüz yok demektir (ilk eşitleme).

export interface SyncProvider {
  /** Şifreli blob'u yükler. Dönen değer: kalıcılaştırılması gereken yeni durum
   *  (örn. ilk push'ta oluşturulan gist id) — yoksa null. */
  upload(content: string): Promise<{ gistId?: string } | null>
  download(): Promise<string | null>
}

const TIMEOUT_MS = 30_000

/** fetch hatalarını FerroError'a çevirir (ağ hatası → CONNECTION_FAILED). */
async function doFetch(url: string, init: RequestInit): Promise<Response> {
  try {
    return await fetch(url, { ...init, signal: AbortSignal.timeout(TIMEOUT_MS) })
  } catch (err) {
    throw new FerroError('CONNECTION_FAILED', 'Sync sunucusuna ulaşılamadı', String(err))
  }
}

function authError(res: Response, what: string): FerroError {
  return new FerroError('AUTH_FAILED', `${what} kimlik doğrulaması başarısız (HTTP ${res.status})`)
}

// ── GitHub Gist ────────────────────────────────────────────────────────────

const GIST_API = 'https://api.github.com/gists'

export class GistProvider implements SyncProvider {
  constructor(
    private readonly token: string,
    private readonly gistId: string
  ) {
    if (!token) throw new FerroError('VALIDATION', 'GitHub token ayarlanmamış')
  }

  private headers(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.token}`,
      Accept: 'application/vnd.github+json',
      'X-GitHub-Api-Version': '2022-11-28',
      'User-Agent': 'ferro-sync'
    }
  }

  async upload(content: string): Promise<{ gistId?: string } | null> {
    const body = JSON.stringify({
      description: 'Ferro sync (encrypted)',
      public: false,
      files: { [SYNC_FILE_NAME]: { content } }
    })
    // Gist id yoksa ilk push: gizli gist oluştur, dönen id kalıcılaştırılır.
    const res = this.gistId
      ? await doFetch(`${GIST_API}/${encodeURIComponent(this.gistId)}`, {
          method: 'PATCH',
          headers: this.headers(),
          body
        })
      : await doFetch(GIST_API, { method: 'POST', headers: this.headers(), body })
    if (res.status === 401 || res.status === 403) throw authError(res, 'GitHub')
    if (res.status === 404 && this.gistId) {
      throw new FerroError('NOT_FOUND', 'Gist bulunamadı — kimliği kontrol edin')
    }
    if (!res.ok) {
      throw new FerroError('CONNECTION_FAILED', `Gist yazılamadı (HTTP ${res.status})`)
    }
    if (!this.gistId) {
      const created = (await res.json()) as { id?: string }
      if (created.id) {
        log.info(`sync gist oluşturuldu: ${created.id}`)
        return { gistId: created.id }
      }
    }
    return null
  }

  async download(): Promise<string | null> {
    if (!this.gistId) return null // ilk eşitleme: uzak kopya henüz yok
    const res = await doFetch(`${GIST_API}/${encodeURIComponent(this.gistId)}`, {
      headers: this.headers()
    })
    if (res.status === 404) return null
    if (res.status === 401 || res.status === 403) throw authError(res, 'GitHub')
    if (!res.ok) {
      throw new FerroError('CONNECTION_FAILED', `Gist okunamadı (HTTP ${res.status})`)
    }
    const gist = (await res.json()) as {
      files?: Record<string, { content?: string; truncated?: boolean; raw_url?: string }>
    }
    const file = gist.files?.[SYNC_FILE_NAME]
    if (!file) return null
    // 1 MiB üstü içerikte API content'i kırpar — ham URL'den çekilir.
    if (file.truncated && file.raw_url) {
      const raw = await doFetch(file.raw_url, { headers: this.headers() })
      if (!raw.ok) {
        throw new FerroError('CONNECTION_FAILED', `Gist içeriği okunamadı (HTTP ${raw.status})`)
      }
      return raw.text()
    }
    return file.content ?? null
  }
}

// ── WebDAV ─────────────────────────────────────────────────────────────────

export class WebdavProvider implements SyncProvider {
  constructor(
    private readonly url: string,
    private readonly user: string,
    private readonly password: string
  ) {
    if (!url) throw new FerroError('VALIDATION', 'WebDAV adresi ayarlanmamış')
  }

  /** Yapılandırılan klasör adresi + sabit dosya adı. */
  private fileUrl(): string {
    return `${this.url.replace(/\/+$/, '')}/${SYNC_FILE_NAME}`
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = {}
    if (this.user || this.password) {
      h.Authorization =
        'Basic ' + Buffer.from(`${this.user}:${this.password}`, 'utf8').toString('base64')
    }
    return h
  }

  async upload(content: string): Promise<{ gistId?: string } | null> {
    const put = (): Promise<Response> =>
      doFetch(this.fileUrl(), {
        method: 'PUT',
        headers: { ...this.headers(), 'Content-Type': 'application/json' },
        body: content
      })
    let res = await put()
    // 409: üst koleksiyon yok — bir kez MKCOL ile oluşturmayı dene.
    if (res.status === 409) {
      await doFetch(this.url.replace(/\/+$/, ''), { method: 'MKCOL', headers: this.headers() })
      res = await put()
    }
    if (res.status === 401 || res.status === 403) throw authError(res, 'WebDAV')
    if (!res.ok) {
      throw new FerroError('CONNECTION_FAILED', `WebDAV'a yazılamadı (HTTP ${res.status})`)
    }
    return null
  }

  async download(): Promise<string | null> {
    const res = await doFetch(this.fileUrl(), { headers: this.headers() })
    if (res.status === 404) return null
    if (res.status === 401 || res.status === 403) throw authError(res, 'WebDAV')
    if (!res.ok) {
      throw new FerroError('CONNECTION_FAILED', `WebDAV'dan okunamadı (HTTP ${res.status})`)
    }
    return res.text()
  }
}

File: src-tauri/src/sync/providers.rs
//! Sync/Team uzak depo sağlayıcıları — `sync/providers.ts` karşılığı (reqwest, async).
//! Yalnızca şifreli blob taşınır. upload → (gist için) gistId döner; download → içerik.

use std::time::Duration;

use serde_json::json;

use crate::error::{FerroError, FerroErrorCode, FerroResult};

const GIST_API: &str = "https://api.github.com/gists";
const TIMEOUT: Duration = Duration::from_secs(30);

pub enum Provider {
    Gist { token: String, gist_id: String, file_name: String },
    Webdav { url: String, user: String, password: String, file_name: String },
}

fn client() -> FerroResult<reqwest::Client> {
    reqwest::Client::builder()
        .timeout(TIMEOUT)
        .build()
        .map_err(|e| FerroError::with_detail(FerroErrorCode::Unknown, "HTTP istemcisi kurulamadı", e.to_string()))
}

fn http_err(e: reqwest::Error, msg: &str) -> FerroError {
    let code = if e.is_timeout() { FerroErrorCode::Timeout } else { FerroErrorCode::ConnectionFailed };
    FerroError::with_detail(code, msg.to_string(), e.to_string())
}

impl Provider {
    /// İçeriği uzağa yazar. Gist için (yeni oluşturulduysa) gistId döner.
    pub async fn upload(&self, content: &str) -> FerroResult<Option<String>> {
        match self {
            Provider::Gist { token, gist_id, file_name } => {
                let c = client()?;
                let body = json!({
                    "description": "Ferro sync (encrypted)",
                    "public": false,
                    "files": { file_name: { "content": content } }
                });
                let req = if gist_id.is_empty() {
                    c.post(GIST_API)
                } else {
                    c.patch(format!("{GIST_API}/{gist_id}"))
                };
                let resp = req
                    .bearer_auth(token)
                    .header("Accept", "application/vnd.github+json")
                    .header("X-GitHub-Api-Version", "2022-11-28")
                    .header("User-Agent", "ferro-sync")
                    .json(&body)
                    .send()
                    .await
                    .map_err(|e| http_err(e, "Gist yüklenemedi"))?;
                check_status(resp.status(), !gist_id.is_empty())?;
                let created: serde_json::Value = resp.json().await.map_err(|e| http_err(e, "Gist yanıtı okunamadı"))?;
                Ok(created.get("id").and_then(|v| v.as_str()).map(|s| s.to_string()))
            }
            Provider::Webdav { url, user, password, file_name } => {
                let c = client()?;
                let file_url = format!("{}/{}", url.trim_end_matches('/'), file_name);
                let put = |c: &reqwest::Client| {
                    let mut r = c.put(&file_url).header("Content-Type", "application/json").body(content.to_string());
                    if !user.is_empty() || !password.is_empty() {
                        r = r.basic_auth(user, Some(password));
                    }
                    r
                };
                let resp = put(&c).send().await.map_err(|e| http_err(e, "WebDAV PUT başarısız"))?;
                if resp.status().as_u16() == 409 {
                    // Koleksiyon yok — oluştur, sonra tekrar dene.
                    let mut mkcol = c.request(reqwest::Method::from_bytes(b"MKCOL").unwrap(), url.trim_end_matches('/'));
                    if !user.is_empty() || !password.is_empty() {
                        mkcol = mkcol.basic_auth(user, Some(password));
                    }
                    let _ = mkcol.send().await;
                    let retry = put(&c).send().await.map_err(|e| http_err(e, "WebDAV PUT (retry) başarısız"))?;
                    check_status(retry.status(), false)?;
                } else {
                    check_status(resp.status(), false)?;
                }
                Ok(None)
            }
        }
    }

    /// Uzak içeriği indirir. Uzak kopya yoksa None.
    pub async fn download(&self) -> FerroResult<Option<String>> {
        match self {
            Provider::Gist { token, gist_id, file_name } => {
                if gist_id.is_empty() {
                    return Ok(None);
                }
                let c = client()?;
                let resp = c
                    .get(format!("{GIST_API}/{gist_id}"))
                    .bearer_auth(token)
                    .header("Accept", "application/vnd.github+json")
                    .header("X-GitHub-Api-Version", "2022-11-28")
                    .header("User-Agent", "ferro-sync")
                    .send()
                    .await
                    .map_err(|e| http_err(e, "Gist indirilemedi"))?;
                if resp.status().as_u16() == 404 {
                    return Ok(None);
                }
                check_status(resp.status(), true)?;
                let gist: serde_json::Value = resp.json().await.map_err(|e| http_err(e, "Gist yanıtı okunamadı"))?;
                let Some(file) = gist.get("files").and_then(|f| f.get(file_name)) else {
                    return Ok(None);
                };
                if file.get("truncated").and_then(|t| t.as_bool()).unwrap_or(false) {
                    if let Some(raw) = file.get("raw_url").and_then(|u| u.as_str()) {
                        let raw_resp = c
                            .get(raw)
                            .bearer_auth(token)
                            .header("User-Agent", "ferro-sync")
                            .send()
                            .await
                            .map_err(|e| http_err(e, "Gist ham içeriği indirilemedi"))?;
                        return Ok(Some(raw_resp.text().await.map_err(|e| http_err(e, "İçerik okunamadı"))?));
                    }
                }
                Ok(file.get("content").and_then(|v| v.as_str()).map(|s| s.to_string()))
            }
            Provider::Webdav { url, user, password, file_name } => {
                let c = client()?;
                let file_url = format!("{}/{}", url.trim_end_matches('/'), file_name);
                let mut r = c.get(&file_url);
                if !user.is_empty() || !password.is_empty() {
                    r = r.basic_auth(user, Some(password));
                }
                let resp = r.send().await.map_err(|e| http_err(e, "WebDAV GET başarısız"))?;
                if resp.status().as_u16() == 404 {
                    return Ok(None);
                }
                check_status(resp.status(), false)?;
                Ok(Some(resp.text().await.map_err(|e| http_err(e, "İçerik okunamadı"))?))
            }
        }
    }
}

fn check_status(status: reqwest::StatusCode, have_id: bool) -> FerroResult<()> {
    let code = status.as_u16();
    if status.is_success() {
        return Ok(());
    }
    if code == 401 || code == 403 {
        return Err(FerroError::new(FerroErrorCode::AuthFailed, "Depo kimlik doğrulaması başarısız (jeton/parola)"));
    }
    if code == 404 && have_id {
        return Err(FerroError::new(FerroErrorCode::NotFound, "Uzak kayıt bulunamadı"));
    }
    Err(FerroError::with_detail(FerroErrorCode::ConnectionFailed, "Depo isteği başarısız", format!("HTTP {code}")))
}
